sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/display/DisplayData.java

```diff
@@ -0,0 +1,376 @@
+   * items.
+   */
+  public static interface ItemBuilder extends Builder {
+    /**
+     * Add a human-readable label to describe the metadata field. A label is optional; if
+     * unspecified, UIs should display the metadata key to identify the display item.
+     */
+    ItemBuilder withLabel(String label);
+
+    /**
+     * Add a link URL to the registered display metadata. A link URL is optional and
+     * can be provided to point the reader to additional details about the metadata.
+     */
+    ItemBuilder withLinkUrl(String url);
+  }
+
+  /**
+   * A display metadata item. DisplayData items are registered via {@link Builder#add} within
+   * {@link HasDisplayData#populateDisplayData} implementations. Each metadata item is uniquely
+   * identified by the specified key and namespace generated from the registering component's
+   * class name.
+   */
+  public static class Item<T> {
+    private final String key;
+    private final String ns;
+    private final Type type;
+    private final T value;
+    private final String label;
+    private final String url;
+
+    private static <T> Item<T> create(String namespace, String key, Type type, T value) {
+      return new Item<>(namespace, key, type, value, null, null);
+    }
+
+    private Item(String namespace, String key, Type type, T value, String url, String label) {
+      this.ns = namespace;
+      this.key = key;
+      this.type = type;
+      this.value = value;
+      this.url = url;
+      this.label = label;
+    }
+
+    String getNamespace() {
+      return ns;
+    }
+
+    Type getType() {
+      return type;
+    }
+
+    T getValue() {
+      return value;
+    }
+
+    String getLabel() {
+      return label;
+    }
+
+    String getUrl() {
+      return url;
+    }
+
+    String getKey() {
+      return key;
+    }
+
+    /**
+     * Retrieve a serialized version of the display item value suitable for use as JSON.
+     */
+    String getValueString() {
+      return type.serializeJsonString(value);
+    }
+
+    private Item<T> addLabel(String label) {
+      return new Item<>(this.ns, this.key, this.type, this.value, this.url, label);
+    }
+
+    private Item<T> addUrl(String url) {
+      return new Item<>(this.ns, this.key, this.type, this.value, url, this.label);
+    }
+  }
+
+  /**
+   * Unique identifier for a display metadata item within a {@link PTransform} instance.
+   * Identifiers are composed of the key the are registered with and a namespace generated from
+   * the class of the component which registered the item.
+   *
+   * <p>Display metadata registered with the same key from different components will have different
+   * namespaces and will thus both be represented in the composed {@link DisplayData}. If a
+   * single component registers multiple metadata items with the same key, only the most recent
+   * item will be retained; previous versions are discarded.
```

Comment:
It may be better to make it an exception to register multiple items with the same key, since otherwise it will lead to unexpected behavior. In the future, we can add the `path` to uniquely identify occurrences in different positions.